### PR TITLE
Source, Obj, Candidate, and Permissions Handler Refactor #2: Move Permissions Checking to Postgres

### PR DESCRIPTION
**Pull request by dannygoldstein**
**High Level Description**

This PR overhauls how database record read/write permissions are enforced in SkyPortal (i.e., the enforcement of who can access / update what data through the SkyPortal API). High level changes introduced here include: 

* Querying of read/write permissions within postgres, rather than python 
* Implementation of common access patterns as dynamically generated mixin classes
* Standardization of access methods across mapped classes
* Introduction of a powerful new accessibility abstraction, mapped class accessibility hybrid methods 
* Ability to filter and join on object accessibility 

***Related PRs***
#1441 

**Motivation and Context**
The notion of who can read a specific database record is fundamental to SkyPortal. As a modern web application with highly granular access permissions across a wide range of database records (Sources, Objects, Photometry, Spectra, etc.), one user's view of the same frontend URL may differ quite drastically from another's due to small differences in access control. As we have seen with, e.g., Photometry, the logic for enforcing who can access which database records can often be quite complex and subtle. An Obj, for example, is accessible to a user only if any of its Photometry is accessible to the User, or if it is saved as a Source or a Candidate to any of the user's accessible groups. To ensure that this logic is implemented consistently and clearly throughout the application, an access checking schema should satisfy the following criteria:

1. It is queryable directly at the database level, e.g., I can look up `SELECT * FROM OBJECTS WHERE OBJECT_IS_READABLE_TO(USER)`
2. It is modular and consistent throughout the application, i.e., there is only one piece of code that checks a user's access to a given type of database record and that piece of code is used consistently throughout the application everywhere where that access check is to be performed
3. Permissions checking is performant, and can be done in batches, even for complex access control logic
4. Permissions checking queries can be combined with other filter clauses and can be used as join criteria on tables
5. There is a common interface across all mapped database classes for permission control queries. 

Currently, 

1. Permissions control is not queryable at the database level, but rather only on a database record-by-database record basis from within python by calling the methods of individual instances of SQLalchemy ORM-mapped classes. As the instantiation of a mapped class pulls over all of its attendant data, this can result in needing to pull over a lot of unneeded data in order to check access control, as well as unnecessary database traffic (many queries sending code and data over the network versus a single batch query). 
2. There are some modular methods of access checking that are used throughout the application (e.g., `get_if_readable_by`), but there are also many examples of the same logic and access checks being implemented directly within API handlers, creating multiple sources of truth for access control logic for a given type of database record (e.g., `.filter(Filter.group_id.in_(group_ids))...` in an API handler) 
3. Permissions control is done on a record-by-record basis, for all types of access. There is no interface for bulk permissions queries.
4. Permissions control is done as an afterburner task to database queries, so the two cannot be combined into a single operation. One cannot currently join on permissions checking.
5. There is not a common interface for access checking across mapped database classes, and methods that do the same thing across multiple classes can have different naming conventions (e.g., `get_if_readable_by`, `get_taxonomy_usable_by_user`, etc.)

This PR refactors the access control infrastructure of skyportal to address these issues.

**Detailed Description**

To address point 1), in this PR I have re-implemented all of the access control queries that were previously done in Python in pure SQL using SQLalchemy core. The first step in this process was to introduce two helper tables, `user_acls_temporary_table` and `user_accessible_groups_temporary_table`. With these helper queries, SQL ownership queries are rendered as correlated subqueries. If I wanted to issue a query for the Object IDs of all objects readable by a given user, first I would construct a join that returns a table containing all (obj_id, user_id) tuples in which the user_id can read the obj_id. I would then construct a table containing *all* user_id, obj_id pairs and OUTER JOIN it against the readable pairs table. A boolean check on whether the readable_pair obj_id is not null after the outer join is exactly what I am after - if this is true, then the object is readable by the user, if not then it is not. By using LATERALs and subquery correlations, these queries take advantage of all available indices and are kept fast.

Using the SQLalchemy `hybrid_method` class, these queries are generated automatically for the user. For example, I could do 

```
objs_i_can_see = Obj.query.filter(Obj.is_readable_by(me))
```

and sqlalchemy will render 

```
SELECT objs.ra, objs.dec, objs.created_at, objs.modified, objs.id, objs.ra_dis, objs.dec_dis, objs.ra_err, objs.dec_err, objs."offset", objs.redshift, objs.redshift_history, objs.altdata, objs.dist_nearest_source, objs.mag_nearest_source, objs.e_mag_nearest_source, objs.transient, objs.varstar, objs.is_roid, objs.score, objs.origin, objs.internal_key, objs.detect_photometry_count
FROM objs
WHERE (SELECT anon_2.cls_id IS NOT NULL AS anon_1
FROM objs AS objs_1 JOIN users AS users_1 ON true LEFT OUTER JOIN LATERAL (SELECT photometry.obj_id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry JOIN group_photometry ON photometry.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT sources.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM sources JOIN user_accessible_groups ON sources.group_id = user_accessible_groups.group_id
WHERE sources.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT candidates.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM candidates JOIN filters ON filters.id = candidates.filter_id JOIN user_accessible_groups ON filters.group_id = user_accessible_groups.group_id
WHERE candidates.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id) AS anon_2 ON objs_1.id = anon_2.cls_id
WHERE objs_1.id = objs.id AND users_1.id = 1) IS true
```

I can also add more conditions, joins, etc.

```
DBSession().query(Photometry).join(Obj).filter(Obj.is_readable_by(u), Photometry.flux > 500)
```

```
SELECT photometry.ra, photometry.dec, photometry.id, photometry.created_at, photometry.modified, photometry.mjd, photometry.flux, photometry.fluxerr, photometry.filter, photometry.ra_unc, photometry.dec_unc, photometry.original_user_data, photometry.altdata, photometry.upload_id, photometry.origin, photometry.obj_id, photometry.instrument_id, photometry.followup_request_id, photometry.assignment_id, photometry.owner_id
FROM photometry JOIN objs ON objs.id = photometry.obj_id
WHERE (SELECT anon_2.cls_id IS NOT NULL AS anon_1
FROM objs AS objs_1 JOIN users AS users_1 ON true LEFT OUTER JOIN LATERAL (SELECT photometry.obj_id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry JOIN group_photometry ON photometry.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT sources.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM sources JOIN user_accessible_groups ON sources.group_id = user_accessible_groups.group_id
WHERE sources.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT candidates.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM candidates JOIN filters ON filters.id = candidates.filter_id JOIN user_accessible_groups ON filters.group_id = user_accessible_groups.group_id
WHERE candidates.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id) AS anon_2 ON objs_1.id = anon_2.cls_id
WHERE objs_1.id = objs.id AND users_1.id = 1) IS true AND photometry.flux > 500
```

I can even join on permissions checking,

```
 DBSession().query(Photometry).join(User, Photometry.is_readable_by(User))
```

gives

```
SELECT photometry.ra, photometry.dec, photometry.id, photometry.created_at, photometry.modified, photometry.mjd, photometry.flux, photometry.fluxerr, photometry.filter, photometry.ra_unc, photometry.dec_unc, photometry.original_user_data, photometry.altdata, photometry.upload_id, photometry.origin, photometry.obj_id, photometry.instrument_id, photometry.followup_request_id, photometry.assignment_id, photometry.owner_id
FROM photometry JOIN users ON (SELECT anon_2.cls_id IS NOT NULL AS anon_1
FROM photometry AS photometry_1 JOIN users AS users_1 ON true LEFT OUTER JOIN LATERAL (SELECT DISTINCT photometry_2.id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry AS photometry_2 JOIN group_photometry ON photometry_2.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry_2.id = photometry_1.id AND user_accessible_groups.user_id = users_1.id) AS anon_2 ON photometry_1.id = anon_2.cls_id
WHERE photometry_1.id = photometry.id AND users_1.id = users.id) IS true
```

These queries allow broadcasting in both the User and `Object` (Photometry, Spectra, ...) directions, enabling them to be used as join and filter clauses. This addresses point 3. The filtering and joining enables batch queries, and reduces the traffic over the network. As a benchmark, the time to execute the permissions checks implemented here for 1000 objects on the production fritz DB  is ~10x faster than what is currently being used. This addresses point 4.

Since the methods are hybrid, they can also be called on individual instances:

```
p = Photometry.query.first()
p.is_readable_by(u)
```

gives

```
True
```

The hybrids can also be queried directly, in addition to being filtered on,

```
DBSession().query(Obj.id, User.id, Obj.is_readable_by(User)).join(User, Obj.is_readable_by(User)).limit(10)
```

emits

```
SELECT objs.id, users.id, (SELECT anon_3.cls_id IS NOT NULL AS anon_2
FROM objs AS objs_1 JOIN users AS users_1 ON true LEFT OUTER JOIN LATERAL (SELECT photometry.obj_id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry JOIN group_photometry ON photometry.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT sources.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM sources JOIN user_accessible_groups ON sources.group_id = user_accessible_groups.group_id
WHERE sources.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT candidates.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM candidates JOIN filters ON filters.id = candidates.filter_id JOIN user_accessible_groups ON filters.group_id = user_accessible_groups.group_id
WHERE candidates.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id) AS anon_3 ON objs_1.id = anon_3.cls_id
WHERE objs_1.id = objs.id AND users_1.id = users.id) IS true AS anon_1
FROM objs JOIN users ON (SELECT anon_5.cls_id IS NOT NULL AS anon_4
FROM objs AS objs_2 JOIN users AS users_2 ON true LEFT OUTER JOIN LATERAL (SELECT photometry.obj_id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry JOIN group_photometry ON photometry.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry.obj_id = objs_2.id AND user_accessible_groups.user_id = users_2.id UNION SELECT sources.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM sources JOIN user_accessible_groups ON sources.group_id = user_accessible_groups.group_id
WHERE sources.obj_id = objs_2.id AND user_accessible_groups.user_id = users_2.id UNION SELECT candidates.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM candidates JOIN filters ON filters.id = candidates.filter_id JOIN user_accessible_groups ON filters.group_id = user_accessible_groups.group_id
WHERE candidates.obj_id = objs_2.id AND user_accessible_groups.user_id = users_2.id) AS anon_5 ON objs_2.id = anon_5.cls_id
WHERE objs_2.id = objs.id AND users_2.id = users.id) IS true
 LIMIT 10
```

Which returns

```
[('14gqr_unsaved_copy', 1, True),
 ('14gqr_unsaved_copy', 2, True),
 ('14gqr_unsaved_copy', 3, True),
 ('14gqr_unsaved_copy', 4, True),
 ('14gqr_unsaved_copy', 5, True),
 ('14gqr', 1, True),
 ('14gqr', 2, True),
 ('14gqr', 3, True),
 ('14gqr', 4, True),
 ('14gqr', 5, True)]
```

The only difference in logic across classes and access types is the generation of the "accessible pairs table," or the table of all (user_id, target_class_id) tuples where user_id can read the database record with PK target_class_id. Once this table is generated, all of the other logic is the same across all class and access types. This in mind, I factored all of the permission control logic into a series of dynamically generated mixin classes that all execute the same underlying code. So if you query for `Spectrum.is_readable_by`, you will be executing the exact same code as `Obj.is_readable_by`, save for the generation of the accessible pair tables for each class. This addresses point 2. 

Because many classes have common access patterns (e.g., this record is only readable if it is accessible to one of the querying user's groups), I consolidated all the access checking into dynamically generated mixin classes that provide a common interface to all permission control logic. 

Classes are generated via opcodes that you specify to a class factory function, `make_permission_control`. For example, I could generate a read-access control mixin class via:

```
ReadProtected = make_permission_control('ReadProtected', 'is_readable')
```

This will create a class that implements `is_readable_by` as a hybrid property, and the classmethod `get_if_is_readable_by`. The class also has two abstract methods that are unimplemented, and the user must implement these if they wish to instantiate the class or any of its subclasses (the pair table generation function and a simple function returning the attributes that subclasses must have in order to successfully mix the class). The methods in these classes call *a single set* of helper methods that is the same across all permission control types, so it is truly the same code that is being executed to check access across all classes.

If I wanted to implement a ReadProtected class that ensures database records are only visible to users that are in any of the object's groups, I would implement a `_readable_pair_table` that joins against the `user_accessible_groups` temporary table and returns the appropriate pairs. By mixing this class in to sqlalchemy mapped classes, they would automatically inherit the logic. 

Modification logic is handled similarly, e.g., I could do

```
WriteProtected = make_permission_control('WriteProtected', 'is_modifiable`)
```

and it would generate a class with methods named of the form `*is_modifiable*`. These could also be mixed into mapped classes, and would not collide with the `is_readable` methods. This addresses point 5.

### CODE REVIEW

File: skyportal/handlers/api/candidate.py [hunk 1]
```diff
@@ -412,10 +423,10 @@ def get(self, obj_id=None):
             "null",
             "undefined",
         ]:
-            start_date = arrow.get(start_date).datetime
+            start_date = arrow.get(start_date).isoformat()
             q = q.filter(Candidate.passed_at >= start_date)
         if end_date is not None and end_date.strip() not in ["", "null", "undefined"]:
-            end_date = arrow.get(end_date).datetime
+            end_date = arrow.get(end_date).isoformat()
```
Comment (dannygoldstein) [inline skyportal/handlers/api/candidate.py:429]: This implementation is more stable for me locally, using datetimes runs into some issues with UTC on my mac OSX.

**Comment by dannygoldstein** [comment]
Failure unrelated

**Comment by dannygoldstein** [comment]
Another cool query: 

How many Obj/User pairs are accessible, and how many are inaccessible?

```

In [13]: q = DBSession().query(Obj.id.label('oid'),  User.id.label('uid'), Obj.is_readable_by(User).label('is_readable')).join(User, sa. literal(True)).subquery()

In [14]: DBSession().query(q.c.is_readable, sa.func.count('*')).group_by(q.c.is_readable).all()
Out[14]: [(False, 4), (True, 20)]
```

**Comment by dannygoldstein** [comment]
How many users can / can't see each object?

```
In [18]: DBSession().query(q.c.is_readable, q.c.oid, sa.func.count('*')).group_by(q.c.is_readable, q.c.oid).all()

SELECT anon_1.is_readable, anon_1.oid, count('*') AS count_1
FROM (SELECT objs.id AS oid, users.id AS uid, (SELECT anon_3.cls_id IS NOT NULL AS anon_2
FROM objs AS objs_1 JOIN users AS users_1 ON true LEFT OUTER JOIN LATERAL (SELECT photometry.obj_id AS cls_id, user_accessible_groups.user_id AS user_id
FROM photometry JOIN group_photometry ON photometry.id = group_photometry.photometr_id JOIN user_accessible_groups ON group_photometry.group_id = user_accessible_groups.group_id
WHERE photometry.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT sources.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM sources JOIN user_accessible_groups ON sources.group_id = user_accessible_groups.group_id
WHERE sources.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id UNION SELECT candidates.obj_id AS obj_id, user_accessible_groups.user_id AS user_id
FROM candidates JOIN filters ON filters.id = candidates.filter_id JOIN user_accessible_groups ON filters.group_id = user_accessible_groups.group_id
WHERE candidates.obj_id = objs_1.id AND user_accessible_groups.user_id = users_1.id) AS anon_3 ON objs_1.id = anon_3.cls_id
WHERE objs_1.id = objs.id AND users_1.id = users.id) IS true AS is_readable
FROM objs JOIN users ON true) AS anon_1 GROUP BY anon_1.is_readable, anon_1.oid

Out[18]:
[(False, '14gqr', 1),
 (False, '14gqr_unsaved_copy', 1),
 (False, '16fil', 1),
 (False, '16fil_unsaved_copy', 1),
 (True, '14gqr', 5),
 (True, '14gqr_unsaved_copy', 5),
 (True, '16fil', 5),
 (True, '16fil_unsaved_copy', 5)]
```

**Comment by acrellin** [comment]
@dannygoldstein In local testing I'm finding that this proposed implementation results in a >10x slowdown for a sample query.

**Comment by dannygoldstein** [comment]
Which query?

________________________________

File: skyportal/models.py [hunk 2]
```diff
@@ -103,55 +211,333 @@ def user_to_dict(self):
...
+    @classmethod
+    def _required_attributes_for_is_readable_check(cls):
+        return ('obj',)
+
+    @classmethod
+    def _is_readable_pair_table(cls, correlation_cls_alias, correlation_user_alias):
+
+        cls_alias = sa.alias(cls)
+        user_alias = sa.alias(User)
+        obj_alias = aliased(Obj)
```
Comment (acrellin) [inline skyportal/models.py:484]: Why `aliased` here, but `sa.alias` elsewhere?

**Review by acrellin — COMMENTED** [review]
Both to aid in reviewing and general accessibility/maintainability, it would be very helpful if you could add both function/class docstrings and inline comments

**Comment by acrellin** [comment]
@kmshin1397 @dannygoldstein  before too much more energy goes into adding to/reviewing this PR, just a heads up that there still needs to be further group discussion around whether this is worth the significant added complexity (@stefanv has expressed reservations/concerns, and has been tied up in a conference but will weigh in tomorrow). @dannygoldstein it doesn't seem like you got everyone's approval on this approach before implementing.

File: skyportal/models.py [hunk 3]
```diff
@@ -103,55 +211,375 @@ def user_to_dict(self):
...
+    pairs = accessible_pair_func(cls.__table__, User.__table__).alias()
+
+    return (
+        DBSession()
+        .query(cls)
+        .join(User, sa.literal(True))
+        .join(pairs, sa.and_(User.id == pairs.c.user_id, cls.id == pairs.c.cls_id))
+        .filter(User.id == accessibility_target)
+        .options(options)
+    )
```
Comment (kmshin1397) [inline skyportal/models.py:252]: @dannygoldstein I'm probably missing something here, but do we really need this pattern of joining the accessible pairs against all possible pairs? Is there a reason we can't just filter directly on the accessible pairs table? Something like `DBSession().query(cls).join(pairs, cls.id == pairs.c.cls_id).filter(pairs.c.user_id == accessibility_target).options(options)`?

Comment (kmshin1397) [reply skyportal/models.py:252]: I just tried this and it seems to return the same results. Only tried it on the one candidates query you and Ari were trying last night for benchmarking though. Very very marginal improvements on the performance as well (like 4-5ms on that query). Could probably do something similar for the `_is_accessible_sql` as well.

**Comment by dannygoldstein** [comment]
Closing in favor of #1542

**Pull request closed without merging**